pyigm/fN/fnmodel.py

```diff
@@ -0,0 +1,456 @@
+          Velocities relative to z
+        cosmo : astropy.cosmology, optional
+
+
+        Returns
+        -------
+
+        log_fNX : float, array, or 2D array
+          f(NHI,X)[z] values
+          Float if given one NHI,z value each. Otherwise 2D array
+          If 2D, it is [NHI,z] on the axes
+
+        """
+        # Tuple?
+        if isinstance(NHI, tuple): # All values packed into NHI parameter
+            z = NHI[1]
+            NHI = NHI[0]
+            flg_1D = 1
+        else:  # NHI and z separate
+            flg_1D = 0
+
+        # NHI
+        if isiterable(NHI):
+            NHI = np.array(NHI)  # Insist on array
+        else:
+            NHI = np.array([NHI])
+        lenNHI = len(NHI)
+
+        # Redshift 
+        if vel_array is not None:
+            z_val = z + (1+z) * vel_array/(const.c.to('km/s').value)
+        else:
+            z_val = z
+        if isiterable(z_val):
+            z_val = np.array(z_val)
+        else:
+            z_val = np.array([z_val])
+        lenz = len(z_val)
+
+        # Check on zmnx
+        bad = np.where((z_val < self.zmnx[0]) | (z_val > self.zmnx[1]))[0]
+        if len(bad) > 0:
+            raise ValueError(
+                'fN.model.eval: z={:g} not within self.zmnx={:g},{:g}'.format(
+                    z_val[bad[0]], *(self.zmnx)))
+
+        if self.mtype == 'Hspline':
+            # Evaluate without z dependence
+            log_fNX = self.model.__call__(NHI)
+
+            # Evaluate
+            if (not isiterable(z_val)) | (flg_1D == 1):  # scalar or 1D array wanted
+                log_fNX += self.gamma * np.log10((1+z_val)/(1+self.zpivot))
+            else:
+                # Matrix algebra to speed things up
+                lgNHI_grid = np.outer(log_fNX, np.ones(len(z_val)))
+                lenfX = len(log_fNX)
+                #
+                z_grid1 = 10**(np.outer(np.ones(lenfX)*self.gamma,
+                                        np.log10(1+z_val)))  #; (1+z)^gamma
+                z_grid2 = np.outer( np.ones(lenfX)*((1./(1+self.zpivot))**self.gamma), 
+                            np.ones(len(z_val)))
+                log_fNX = lgNHI_grid + np.log10(z_grid1*z_grid2) 
+
+        # Gamma function (e.g. Inoue+14)
+        elif self.mtype == 'Gamma':
+            # Setup the parameters
+            Nl, Nu, Nc, bval = self.param[0]
+
+            # gNHI
+            Bi = self.param[1]
+            ncomp = len(Bi)
+            log_gN = np.zeros((lenNHI, ncomp))
+            beta = [item[1] for item in self.param[2:]] 
+            for kk in range(ncomp):
+                log_gN[:, kk] += (np.log10(Bi[kk]) + NHI*(-1 * beta[kk])
+                                + (-1. * 10.**(NHI-Nc) / np.log(10)))  # log10 [ exp(-NHI/Nc) ]
+            # f(z)
+            fz = np.zeros((lenz, 2))
+            # Loop on NHI
+            for kk in range(ncomp):
+                if kk == 0:  # LyaF
+                    zcuts = self.param[2][2:4]
```

Comment:
I guess this works ok for now, but using dictionary for param lists in the future may improve readability and avoid subtle bugs.